### DISCUSSION TITLE
To async or not to async?

### DISCUSSION THREAD

**Discussion by floatingpurr** [fastapi/fastapi#5227] [Questions] [answered]
In all the examples from the doc, I see path operation functions using `async` even if they do not contain I/O-bound operations or calls to third-party libraries with `await`. E.g.,:

```python
@app.get("/")
async def root():
    return {"message": "Hello World"}
```

Why not just use `def`?

The FastAPI doc [states](https://fastapi.tiangolo.com/async/#in-a-hurry):

> If you are using third party libraries that tell you to call them with `await`, then declare your path operation functions with `async def`

Those functions are run in an external threadpool

> If you are using a third party library that communicates with something (a database, an API, the file system, etc) and doesn't have support for using await, (this is currently the case for most database libraries), then declare your path operation functions as normally, with just `def`

Those functions are executed in the event loop

> If your application (somehow) doesn't have to communicate with anything else and wait for it to respond, use async def.

Those functions, like the one in the example, are executed in the event loop too. Is there any advantage in doing that?

**Comment by YuriiMotov** [comment, marked as answer]
If you use endpoint function or dependency declared with `def` (not `async def`), FastAPI will execute them in thread pool.
This brings some overheads. If you are sure your endpoint is light-weight (don't contain heavy computations or other sync blocking code), it't better to make this function `async`. This way it will be executed in event loop and avoid those overheads with using thread pool.
